library/Booster/Pattern/ApplyEquations.hs

```diff
@@ -688,27 +688,64 @@ simplifyConstraints doTracing def mbApi cache ps =
     runEquationT doTracing def mbApi cache $ mapM (simplifyConstraint' True) ps
 
 -- version for internal nested evaluation
-simplifyConstraint' :: MonadLoggerIO io => Bool -> Predicate -> EquationT io Predicate
+simplifyConstraint' :: MonadLoggerIO io => Bool -> Term -> EquationT io Term
 -- We are assuming all predicates are of the form 'true \equals P' and
 -- evaluating them using simplifyBool if they are concrete.
 -- Non-concrete \equals predicates are simplified using evaluateTerm.
-simplifyConstraint' recurse = \case
-    Predicate t@(Term TermAttributes{canBeEvaluated} _)
+simplifyConstraint' recurseIntoEvalBool = \case
+    t@(Term TermAttributes{canBeEvaluated} _)
         | isConcrete t && canBeEvaluated -> do
             mbApi <- (.llvmApi) <$> getConfig
             case mbApi of
                 Just api ->
-                    if simplifyBool api t
-                        then pure $ Predicate TrueBool
-                        else pure $ Predicate FalseBool
-                Nothing ->
-                    Predicate <$> if recurse then evalBool t else pure t
-        | otherwise ->
-            Predicate <$> if recurse then evalBool t else pure t
+                    pure $
+                        if simplifyBool api t
+                            then TrueBool
+                            else FalseBool
+                Nothing -> if recurseIntoEvalBool then evalBool t else pure t
+    NotBool x -> negateBool <$> recursion x
+    EqualsK (KSeq _ l) (KSeq _ r) -> evalEqualsK l r
+    NEqualsK (KSeq _ l) (KSeq _ r) -> negateBool <$> evalEqualsK l r
+    t -> if recurseIntoEvalBool then evalBool t else pure t
   where
     evalBool :: MonadLoggerIO io => Term -> EquationT io Term
     evalBool t = do
         prior <- getState -- save prior state so we can revert
         result <- iterateEquations 100 TopDown PreferFunctions t
         EquationT $ lift $ lift $ put prior
         pure result
+
+    recursion = simplifyConstraint' recurseIntoEvalBool
+
+    negateBool TrueBool = FalseBool
+    negateBool FalseBool = TrueBool
+    negateBool x = NotBool x
+
+    foldAndBool [] = TrueBool
+    foldAndBool [x] = x
+    foldAndBool (FalseBool : _) = FalseBool
+    foldAndBool (TrueBool : xs) = foldAndBool xs
+    foldAndBool (x : xs) = AndBool x $ foldAndBool xs
```

Comment:
This could probably go into said `Booster.Pattern.Bools` module if we end up splitting off those domain modules.

Reply:
Done